### PR TITLE
Add sync module to rustc_data_structures

### PR DESCRIPTION
**Pull request by Zoxc**
This PR is split out from https://github.com/rust-lang/rust/pull/45912, since github apparently can't handle such large PRs.

r? @arielb1

### CODE REVIEW

File: src/librustc/Cargo.toml [hunk 1]
```diff
@@ -15,7 +15,7 @@ fmt_macros = { path = "../libfmt_macros" }
 graphviz = { path = "../libgraphviz" }
 jobserver = "0.1"
 log = "0.3"
-owning_ref = "0.3.3"
+owning_ref = { git = "https://github.com/Zoxc/owning-ref-rs.git" }
```
Comment (arielb1) [inline src/librustc/Cargo.toml]: the rustc build process shouldn't depend on random git repos - that's just too fragile. We want to keep things building in N years.

**Comment by arielb1** [comment]
r=me if you (temporarily) vendor your version of `owning_ref` into the rustc repo

**Comment by Zoxc** [comment]
I added my `owning_ref` fork to a `rustc_data_structures` module.

**Comment by kennytm** [comment]
Please rebase on latest master to pick up the newest toolstate.

**Pull request merged**